util/util.go

```diff
@@ -303,35 +303,59 @@ func IsEmptyExpr(expr ast.Expr) bool {
 	return false
 }
 
-// TypeIsErrorType checks if the type is an error type
-func TypeIsErrorType(typ types.Type) bool {
-	if typ, ok := typ.(*types.Named); ok {
-		return typ.String() == "error"
+const (
+	_errorType = "error"
+	_boolType  = "bool"
+)
+
+// TypeIsRichCheckEffectType checks if the type is an `error` type or an ok form `bool` type.
+func TypeIsRichCheckEffectType(typ types.Type, typName string) bool {
+	switch t := typ.(type) {
+	case *types.Named:
+		return t.String() == typName
+	case *types.Basic:
+		return t.Kind() == types.Bool && t.String() == typName
 	}
 	return false
 }
 
-// FuncIsErrReturning encodes the conditions that a function is deemed "error-returning"
-// this guards its results to require an `err` check before use as nonnil.
-// a function is deemed "error-returning" iff it has a single result of type `error`, and that
-// result is the last in the list of results.
-func FuncIsErrReturning(fdecl *types.Func) bool {
+// funcIsRichCheckEffectReturning encodes the conditions that a function is deemed "rich-check-effect-returning", i.e.,
+// it is an error-returning function or a bool(ok)-returning function.
+// A function is deemed "rich-check-effect-returning" iff it has a single result of type `typName` (error or bool),
+// and that result is the last in the list of results.
+func funcIsRichCheckEffectReturning(fdecl *types.Func, typName string) bool {
 	results := fdecl.Type().(*types.Signature).Results()
 	n := results.Len()
 	if n == 0 {
 		return false
 	}
-	if !TypeIsErrorType(results.At(n - 1).Type()) {
+	if !TypeIsRichCheckEffectType(results.At(n-1).Type(), typName) {
```

Comment:
It's safer to not check against strings. 

See https://github.com/uber-go/nilaway/blob/main/util/util.go#L31, we can look up the Universe scope to get the builtin types and then directly compare against that (speaking of which, we should probably move that to `util/typeshelper` as well).

This would prevent us (again, unlikely to happen in production code) from misunderstanding a shadowed type, e.g., nothing prevents devs to do

```go
type error int
```

(perhaps let's add a test case for that)

Reply:
Ah, that's an interesting pointer. Implemented this safeguard and added test cases under `contracts/namedtypes`. (To avoid cluttering this PR, I'll move to `util/typeshelper` in a follow-up PR.)